File: src/main/java/io/github/svegon/capi/event/render/tooltip/GetItemStackTooltipCallback.java
package io.github.svegon.capi.event.render.tooltip;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.item.TooltipContext;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import org.jetbrains.annotations.Nullable;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

import java.util.List;

@FunctionalInterface
public interface GetItemStackTooltipCallback {
    Event<GetItemStackTooltipCallback> EVENT = EventFactory.createArrayBacked(GetItemStackTooltipCallback.class,
            (stack, player, context, callback) -> {}, listeners -> (stack, player, context, callback) -> {
        for (GetItemStackTooltipCallback listener : listeners) {
            listener.getItemStackTooltip(stack, player, context, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void getItemStackTooltip(ItemStack stack, @Nullable PlayerEntity player, TooltipContext context,
                             CallbackInfoReturnable<List<Text>> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/tooltip/GetItemStackTooltipCallback.kt
package io.github.svegon.capi.event.render.tooltip

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.item.TooltipContext
import net.minecraft.text.Text

fun interface GetItemStackTooltipCallback {
    fun getItemStackTooltip(
        stack: ItemStack?, player: PlayerEntity?, context: TooltipContext?,
        callback: CallbackInfoReturnable<List<Text?>?>?
    )

    companion object {
        @JvmField
        val EVENT: Event<GetItemStackTooltipCallback> = EventFactory.createArrayBacked<GetItemStackTooltipCallback>(
            GetItemStackTooltipCallback::class.java,
            GetItemStackTooltipCallback { stack: ItemStack?, player: PlayerEntity?, context: TooltipContext?, callback: CallbackInfoReturnable<List<Text?>?>? -> }) { listeners: Array<GetItemStackTooltipCallback> ->
            GetItemStackTooltipCallback { stack: ItemStack?, player: PlayerEntity?, context: TooltipContext?, callback: CallbackInfoReturnable<List<Text?>?> ->
                for (listener in listeners) {
                    listener.getItemStackTooltip(stack, player, context, callback)

                    if (callback.isCancelled()) {
                        return@GetItemStackTooltipCallback
                    }
                }
            }
        }
    }
}
